src/mscorlib/src/System/Utf8String.cs

```diff
@@ -0,0 +1,26 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.CompilerServices;
+
+namespace System
+{
+    // This is an experimental type and not referenced from CoreFx but needs to exists and be public so we can prototype in CoreFxLab.
+    public sealed class Utf8String
```

Comment:
As we add interfaces (`IEnumerable, IComparable, ...`) to this during development, will we need to touch the VM code at all? That is, does the VM need to know about every possible member that sits on this type?

Reply:
Not necessarily about interfaces. But the VM and JIT will certainly need to know more about this type to get decent performance. It goes back to my question on what you expect to get by adding this stub to CoreLib.

Reply:
@jkotas You're right in that we're not going to see full performance benchmarks until JIT support comes online, but this does get us at least partway there. Even when not JIT-optimized, the supplied allocation routine will still be faster and allocate less overall heap memory than calling two different ctors for a container + separate array.

Reply:
Could just be a `struct` wrapper with array as its single field?

Reply:
@benaadams We considered this for prototype purposes. But the end result is to have the final type be a class, and any data we'd collect in the meantime from a struct stand-in would be suspect.

Reply:
> the supplied allocation routine will still be faster

That is not correct. The allocation routine in this PR is super slow naive implementation. I just tried:

```
class MyUtf8String
{
   byte[] _payload;

   static public MyUtf8String FastAllocate(int length) { var ret = new MyUtf8String(); ret._payload = new byte[length+1]; return ret; }
}
```

Run `Utf8String.FastAllocate` vs. `MyUtf8String.FastAllocate` in a loop. MyUtf8String loop is faster.

I am with Stephen that it would be best for a rough early prototype code like this to stay in a branch and not be merged into master.

FWIW, `Span<T>` was also developed in a branch and only integrated to master once it was sufficiently along and we agreed to ship it for real: #7886.